File: src/Solutions/Euler662.hs
module Solutions.Euler662 where
import Control.Monad
import Control.Monad.ST
import Data.Array.Unboxed
import Data.Array.ST
import Data.Word
import Utils.Array(modifyArray)
import Utils.Numeric(fibonacci)

type W = Word32

squares :: W -> [(W, W)]
squares n = [(x, y) | x <- [0..n], y <- [0..n], x^2 + y^2 == n^2]

m = 1000000007 :: W

plus :: W -> W -> W
plus a b = let s = a + b in if s > m then s - m else s

funF :: W -> W -> W
funF width height =
    let paths = concatMap squares $ takeWhile (<width + height) $ fibonacci 1 2
    in runST $ do
        arr <- newArray ((0, 0), (width, height)) 0 :: ST s (STUArray s (W, W) W)
        writeArray arr (0, 0) 1
        forM_ paths $ \(dx, dy) -> forM_ [0..width] $ \x -> forM_ [0..height] $ \y ->
            when (x >= dx && y >= dy) $ readArray arr (x - dx, y - dy) >>= modifyArray arr (x, y) . plus
        readArray arr (width, height)

euler662 :: IO String
euler662 = return $ show $ funF 10000 10000

File: rust/src/solutions/euler662.rs
use crate::utils::numeric::sequences::fibonacci::Fibonacci;
use num::integer::Roots;

fn squares(n: usize) -> impl Iterator<Item = (usize, usize)> {
    (0..=n).filter_map(move |x| {
        let y = (n * n - x * x).sqrt();
        if x * x + y * y == n * n {
            Some((x, y))
        } else {
            None
        }
    })
}

fn plus(a: &mut u32, b: u32) {
    let p = 1_000_000_007;
    *a += b;
    if *a >= p {
        *a -= p
    }
}

fn f(width: usize, height: usize) -> u32 {
    let paths = Fibonacci::new(1, 2)
        .take_while(|fib| fib < &(width + height))
        .flat_map(squares)
        .collect::<Vec<_>>();
    let mut arr = vec![vec![0u32; height + 1]; width + 1];
    arr[0][0] = 1;
    for x in 0..=width {
        for y in 0..=height {
            for &(dx, dy) in &paths {
                if x >= dx && y >= dy {
                    let b = arr[x - dx][y - dy];
                    plus(arr.get_mut(x).unwrap().get_mut(y).unwrap(), b);
                }
            }
        }
    }
    arr[width][height]
}

pub fn euler662() -> String {
    f(10_000, 10_000).to_string()
}
